### PR TITLE
Feature/377 constraints and expectations

### PR DESCRIPTION
**Pull request by zzeekk**
### What changes are included in the pull request?
see #377, #388

### CODE REVIEW

**Comment by zzeekk** [comment]
To be discussed in our weekly...
Missing point from my side:
- create a docu site for data quality?
- check integration of original spark metrics.

**Comment by zzeekk** [comment]
> * create a docu site for data quality?
-> done
> * check integration of original spark metrics.
-> looks good

@pgruetter : PR ok for you?

File: sdl-core/src/main/scala/io/smartdatalake/workflow/dataobject/Constraint.scala [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+/**
+ * Definition of row-level constraint to validate.
+ *
+ * @param name name of the constraint
+ * @param description optional detailed description of the constraint
+ * @param expression SQL expression to evaluate on every row. The expressions return value should be a boolean.
+ *                   If it evaluates to true the constraint is validated successfully, otherwise it will throw an exception.
+ * @param errorMsgCols Optional list of column names to add to error message.
+ *                     Note that primary key colums are always included.
```
Comment (pgruetter) [inline sdl-core/src/main/scala/io/smartdatalake/workflow/dataobject/Constraint.scala]:
```suggestion
 *                     Note that primary key columns are always included.
```

**Review by pgruetter — APPROVED** [review]
LGTM, although I only skimmed some parts..
I think this warrants a separate Reference page (or even accompanying blog article), maybe create an issue as reminder?

**Comment by zzeekk** [comment]
Yes, docu site is already existing, but description of expectations must be completed:
http://smartdatalake.ch/docs/reference/dataQuality

**Pull request merged**